DIPs/1NNN-AP.md

```diff
@@ -0,0 +1,181 @@
+| Status:         | Draft                                                           |
+
+## Abstract
+
+Enum Type Inference (ETI) is a shortcut to allow the omission of an enum member's
+type name when it can be contextually inferred.
+
+While D's pre-existing syntax makes ETI a more complex feature to implement than in brand
+new languages, its presence in many new languages has proven its convenience and popularity.
+
+## Contents
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+* [Grammatical Changes](#grammatical-changes)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+Writing the same enum type names over and over again (e.g. exhaustive enum switch-cases)
+involves the unnecessary, often gratuitous repetition of the enum member's type name.
+
+The solution used by many other modern languages is simple: permit the omission
+of the enum member's type name when it can be inferred from its context.
+Dlang having this shortcut would be equally beneficial, with few drawbacks.
+
+## Prior Work
+Implementation of this feature in other languages:
+- [Swift](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html)
+- [Ziglang](https://ziglang.org/documentation/master/#Enum-Literals)
+- [Odin](https://odin-lang.org/docs/overview/#implicit-selector-expression)
+
+Java allows omitting an enum type in switch-case statements: [https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html)
+
+A suggestion for a similar feature in Rust, but with a different approach: [https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851](https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851)
+
+## Description
+Due to `.enumMember` being the same syntax as the [module scope operator](https://dlang.org/spec/module.html#module_scope_operators),
+I propose that ETI should use the syntax `$enumMember`. I'm open to discussing alternatives,
+but I believe that my proposed syntax represents the best compromise between convenience and compatibility.
+
+#### 1. Assignment statements.
+Assigning to a variable with a known type should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+struct B{ A one, two; }
+
+void main(){
+    A    myA1 = $b;      //myA1 = A.b
+    A    myA2 = $b | $c; //myA2 = A.c
+    auto myA3 = $b;      //error, we don't know the type of "$b"!
+    
+    B myB;
+    myB.one = $c; //myB.one = A.c
+    myB.two = $d; //myB.two = A.d
+}
+```
+
+#### 2. Return statements.
+If a function has an explicit enum return type, its return statement(s)
+should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+A myFunc(){
+    return $c; //returns A.c
+}
+
+auto myBrokenFunc(){
+    return $c; //error, we don't know the type of "$c"!
+}
+```
+
+#### 3. Parameters.
+Any explicitly typed enum parameters should allow ETI.
```

Comment:
Just to mention, this could include default arguments to avoid repeating the enum type:
```d
void f(A e = $b);
```

Reply:
Good suggestion! ;)